Add proposal for low-level atomics (#1142)
apple/swift-evolution: proposals/NNNN-atomics.md

We expect code using these unsafe precursors will be easily upgradeable to their eventual non-copiable variants when it becomes possible to implement those. In the meantime, since these are memory-unsafe variants of eventual `Atomic` and `AtomicLazyReference` types, it seems appropriate to simply prefix their names with the customary `Unsafe` prefix.

Logically, `UnsafeAtomic` and `UnsafeAtomicLazyReference` are both reference types with an independent storage representation and manual memory management. The common set of APIs between these types establishes a new pattern geared specifically for modeling low-overhead synchronization constructs in current versions of Swift:

```swift
struct UnsafeDemo {
  typealias Value

  struct Storage {
    // Initialize a new storage instance by converting the given value.
    // The conversion may involve side effects such as unbalanced retain/release
    // operations; to ensure correct results, the resulting storage instance

>     // must be used to initialize exactly once to initialize a memory location.

Author: Ah, this is garbled, what I want to say is

```suggestion
    // must be used to initialize exactly one memory location.
```

The `weakCompareExchange` form may sometimes return false even when the original and expected values are equal. (Such failures may happen when some transient condition prevents the underlying operation from succeeding -- such as an incoming interrupt during a load-link/store-conditional instruction sequence.) This variant is designed to be called in a loop that only exits when the exchange is successful; in such loops using `weakCompareExchange` may lead to a performance improvement by eliminating a nested loop in the regular, "strong", `compareExchange` variants.

The compare-exchange primitive is special: it is a universal operation that can be used to implement all other atomic operations, and more. For example, here is how we could use `compareExchange` to implement a wrapping increment operation over `UnsafeAtomic<Int>` values:

```swift
extension UnsafeAtomic where Value == Int {
  func wrappingIncrement(
    by operand: Int,
    ordering: AtomicUpdateOrdering

>   ) -> Int {

Reviewer: No return value?

```suggestion
  ) {
```

Author: This is quite intentional. There are separate variants for increments/decrements with a return value.

Reviewer: So, should the `-> Int` be removed for this variant?

Author: 🤦‍♂️

| Method Name | Returns | Implements |
| --- | --- | --- |
| `loadThenWrappingIncrement(by:ordering:)`  | original value | `a &+= b`  |
| `loadThenWrappingDecrement(by:ordering:)`  | original value | `a &-= b`  |
| `loadThenBitwiseAnd(with:ordering)`        | original value | `a &= b`  |
| `loadThenBitwiseOr(with:ordering)`         | original value | `a \|= b`  |
| `loadThenBitwiseXor(with:ordering)`        | original value | `a ^= b`   |
| `wrappingIncrementThenLoad(by:ordering:)`  | new value  | `a &+= b`  |
| `wrappingDecrementThenLoad(by:ordering:)`  | new value  |`a &-= b`   |
| `bitwiseAndThenLoad(with:ordering)`        | new value  |`a &= b`    |
| `bitwiseOrThenLoad(with:ordering)`         | new value  |`a \|= b`   |
| `bitwiseXorThenLoad(with:ordering)`        | new value  |`a ^= b`    |
| `wrappingIncrement(by:ordering:)`          | none   | `a &+= b` |
| `wrappingDecrement(by:ordering:)`          | none   | `a &-= b` |

The `wrappingIncrement` and `wrappingDecrement` operations are provided as a convenience for incrementing/decrementing values in the common case when a return value is not required.

> While we require all atomic operations to be free of locks, we don't require wait-freedom. Therefore, on architectures that don't provide direct hardware support some or all of these operations, we still require them to be implemented using `compareExchange` loops like the one for `wrappingIncrement` above.

Reviewer: Missing "for"?

```suggestion
While we require all atomic operations to be free of locks, we don't require wait-freedom. Therefore, on architectures that don't provide direct hardware support for some or all of these operations, we still require them to be implemented using `compareExchange` loops like the one for `wrappingIncrement` above.
```

Note that loading the atomic unmanaged reference and converting it to a strong reference are two distinct operations that won't execute as a single atomic transaction. This can easily lead to race conditions when a thread releases an object while another is busy loading it:

```swift
// BROKEN CODE. DO NOT EMULATE IN PRODUCTION.
let myAtomicRef = UnsafeAtomic<Unmanaged<Foo>>.create(initialValue: ...)

// Thread A: Load the unmanaged value and then convert it to a regular
//           strong reference.
let ref = myAtomicRef.load(ordering: .acquiring).takeUnretainedValue()
...

// Thread B: Store a new reference in the atomic unmanaged value and 
//           release the previous reference.
let new = Unmanaged.passRetained(...)
let old = myAtomicRef.exchange(new, ordering: .acquiringAndReleasing)
old.release() // RACE CONDITION
```

If thread B happens to release the same object that thread A is in the process of loading, then thread A's `takeUnretainedValue` may attempt to retain a deallocated object.

> Such problems make `AtomicUnmanaged` exceedingly difficult to use in all but the simplest situations. The section on [*Atomic Strong References*](#atomic-strong-references-and-the-problem-of-memory-reclamation) below describes some new constructs we may introduce in future proposals to assist with this issue.

Reviewer: `AtomicUnmanaged` ➡️ `UnsafeAtomic<Unmanaged<T>>`

```suggestion
Such problems make `UnsafeAtomic<Unmanaged<T>>` exceedingly difficult to use in all but the simplest situations. The section on [*Atomic Strong References*](#atomic-strong-references-and-the-problem-of-memory-reclamation) below describes some new constructs we may introduce in future proposals to assist with this issue.
```

> For now, we provide the standalone type `UnsafeAtomicLazyReference`; this is an example of a useful construct that could be built on top of `UnsafeAtomic<Unmanaged>` operations. Of all the various atomic constructs introduced in this proposal, only `UnsafeAtomicLazyReference` represents a regular strong reference to a class instance -- the other pointer/reference types leave memory management entirely up to the user.)

Reviewer: Add or remove parenthesis?

```suggestion
For now, we provide the standalone type `UnsafeAtomicLazyReference`; this is an example of a useful construct that could be built on top of `UnsafeAtomic<Unmanaged>` operations. (Of all the various atomic constructs introduced in this proposal, only `UnsafeAtomicLazyReference` represents a regular strong reference to a class instance -- the other pointer/reference types leave memory management entirely up to the user.)
```

> An `UnsafeAtomicLazyReference` holds an optional reference that is initially set to `nil`. The value can be set exactly once, but it can be read an arbitrary number of times. Attempts to change the value after the first `storeIfNilThenLoad` call are ignored, and return the current value instead.

Reviewer: `storeIfNilThenLoad` ➡️ `storeIfNil` to match implementation?

```suggestion
An `UnsafeAtomicLazyReference` holds an optional reference that is initially set to `nil`. The value can be set exactly once, but it can be read an arbitrary number of times. Attempts to change the value after the first `storeIfNil` call are ignored, and return the current value instead.
```

Author: Ah, good eye -- I'm actually changing the implementation to match.

```swift
public struct UnsafeAtomicLazyReference<Instance: AnyObject> {
  public typealias Value = Instance?

  public struct Storage {
    public init()

    @discardableResult 
    public mutating func dispose() -> Value
  }

  public init(at address: UnsafeMutablePointer<Storage>)

  public static func create() -> Self
  @discardableResult 
  public func destroy() -> Value

>   public func storeIfNilThenLoad(_ desired: __owned Instance) -> Instance

Reviewer: ```suggestion
  public func storeIfNil(_ desired: __owned Instance) -> Instance
```

  public func load() -> Instance?
}
```

This is the only atomic type in this proposal that doesn't provide the usual `load`/`store`/`exchange`/`compareExchange` operations.

This construct allows library authors to implement a thread-safe lazy initialization pattern:

```swift
var _foo: UnsafeAtomicLazyReference<Foo> = ...

// This is safe to call concurrently from multiple threads.
var atomicLazyFoo: Foo {
  if let foo = _foo.load() { return foo }
  // Note: the code here may run concurrently on multiple threads.
  // All but one of the resulting values will be discarded.
  let foo = Foo()

>   return _foo.storeIfNilThenLoad(foo)

Reviewer: ```suggestion
  return _foo.storeIfNil(foo)
```

}
```

The Standard Library has been internally using such a pattern to implement deferred bridging for `Array`, `Dictionary` and `Set`.

> Note that unlike the rest of the atomic types, `load` and `storeIfNilThenLoad(_:)` do not expose `ordering` parameters. (Internally, they map to acquiring/releasing operations to guarantee correct synchronization.)

Reviewer: ```suggestion
Note that unlike the rest of the atomic types, `load` and `storeIfNil(_:)` do not expose `ordering` parameters. (Internally, they map to acquiring/releasing operations to guarantee correct synchronization.)
```

We consider these ordering arguments to be an essential part of low-level atomic operations, and we require an explicit `ordering` argument on all atomic operations provided by `UnsafeAtomic`. The intention here is to force developers to carefully think about what ordering they need to use, each time they use one of these primitives. (Perhaps more importantly, this also makes it obvious to readers of the code what ordering is used -- making it far less likely that an unintended `.sequentiallyConsistent` ordering slips through code review.) 

Projects that prefer to default to sequentially consistent ordering are welcome to add non-public `UnsafeAtomic` extensions that implement that. However, we expect that providing an implicit default ordering would be highly undesirable in most production uses of atomics.

Atomic orderings are grouped into three frozen structs based on the kind of operation to which they are attached, as listed below. By modeling these as separate types, we can ensure that unsupported operation/ordering combinations (such as an atomic "releasing load") will lead to clear compile-time errors:

```swift

> @frozen
> struct AtomicLoadOrdering {
>   static var relaxed: AtomicLoadOrdering { get }
>   static var acquiring: AtomicLoadOrdering { get }
>   static var sequentiallyConsistent: AtomicLoadOrdering { get }
> }
> 
> @frozen
> struct AtomicStoreOrdering {
>   static var relaxed: AtomicStoreOrdering { get }
>   static var releasing: AtomicStoreOrdering { get }
>   static var sequentiallyConsistent: AtomicStoreOrdering { get }
> }
> 
> @frozen
> struct AtomicUpdateOrdering {
>   static var relaxed: AtomicUpdateOrdering { get }
>   static var acquiring: AtomicUpdateOrdering { get }
>   static var releasing: AtomicUpdateOrdering { get }
>   static var acquiringAndReleasing: AtomicUpdateOrdering { get }
>   static var sequentiallyConsistent: AtomicUpdateOrdering { get }
> }

Reviewer: Use `Self` to match implementation?

```suggestion
@frozen
struct AtomicLoadOrdering {
  static var relaxed: Self { get }
  static var acquiring: Self { get }
  static var sequentiallyConsistent: Self { get }
}

@frozen
struct AtomicStoreOrdering {
  static var relaxed: Self { get }
  static var releasing: Self { get }
  static var sequentiallyConsistent: Self { get }
}

@frozen
struct AtomicUpdateOrdering {
  static var relaxed: Self { get }
  static var acquiring: Self { get }
  static var releasing: Self { get }
  static var acquiringAndReleasing: Self { get }
  static var sequentiallyConsistent: Self { get }
}
```

Luckily, configuring these special functions to always get force-inlined into all callers guarantees that constant folding will get rid of the switch statement *as long as the supplied ordering is a compile-time constant*. Unfortunately, it's all too easy to accidentally violate this latter requirement, with dire consequences to the expected performance of the atomic operation.

Consider the following well-meaning attempt at using `compareExchange` to define an atomic integer addition operation that traps on overflow rather than allowing the result to wrap around:

```swift
extension UnsafeAtomic where Value == Int {
  // Non-inlinable
  public func checkedIncrement(by delta: Int, ordering: AtomicUpdateOrdering) {
    var done = false
    var current = load(ordering: .relaxed)
    while !done {
      (done, current) = compareExchange(
        expected: current,
        desired: current + operand, // Traps on overflow
        ordering: ordering)
    }
  }
}

// Elsewhere:
counter.checkedIncrement(by: 1, ordering: .relaxed)
```

> If for whatever reason the Swift compiler isn't able (or willing) to inline the `checkedIncrement` call, then the value of `ordering` won't be known at compile time to the body of the function, so even though `compileExchange` will still get inlined, its switch statement won't be eliminated. This leads to a potentially significant performance regression that could interfere with the scalability of the operation.

Reviewer: `compileExchange` ➡️ `compareExchange`

```suggestion
If for whatever reason the Swift compiler isn't able (or willing) to inline the `checkedIncrement` call, then the value of `ordering` won't be known at compile time to the body of the function, so even though `compareExchange` will still get inlined, its switch statement won't be eliminated. This leads to a potentially significant performance regression that could interfere with the scalability of the operation.
```

    _mutex.unlock()
  }
}

var i: Int
...
i = 0
DispatchQueue.concurrentPerform(iterations: 10) { _ in
  for _ in 0 ..< 1_000_000 {
    i.atomicIncrement()  // Exclusivity violation
  }
}
print(i)
```

Even though `NSLock` does guarantee that the `self += 1` line is always serialized, the concurrent `atomicIncrement` invocations still count as an exclusivity violation, because the write access to `i` starts when the function call begins, before the call to `_mutex.lock()`. Therefore, the code above has undefined behavior, despite all the locking. (For example, it may print any value between one and ten million, or it may trap in a runtime exclusivity check, or indeed it may do something else.)

Note that this restriction wasn't introduced by our new low-level atomic primitives -- it is a preexisting property of the language.

This is one of the reasons why `NaïveFuture` was declared as a class above. Class instance methods are allowed to mutate their stored properties without declaring themselves `mutating`, and thus they are outside the scope of the Law of Exclusivity. (Of course, their implementation must still guarantee that the Law is upheld for any variables they access.)

> > **Note:** A more fundamental reason why `NaïveFuture` is a class is that synchronization constructs are difficult to model with value types -- their instances tend to have an inherent identity that prevents copies from working like the original, they often need to be backed by a stable memory location, etc. The [Ownership Manifesto]'s non-copiable types may eventually provide a more efficient and safer model for such constructs, but in today's Swift, we need to represent them with some reference type instead: typically, either a class (like `NaïveFuture`) or some unsafe pointer type (like `UnsafeAtomic`).

Reviewer: The `NaïveFuture` class was removed (in version `2020-03-30/2`).

```swift
class BrokenAtomicCounter { // THIS IS BROKEN; DO NOT USE
  private var _storage = UnsafeAtomic<Int>.Storage(0)
  private var _value: UnsafeAtomic<Int>?
      
  init() {
    // This escapes the ephemeral pointer generated by the inout expression,
    // so it leads to undefined behavior when the pointer gets dereferenced
    // in the atomic operations below. DO NOT DO THIS.
    _value = UnsafeAtomic<Int>(at: &_storage)
  }
  
  func increment() {
    _value!.wrappingIncrement(by: 1, ordering: .relaxed)
  }

  func get() -> Int {
    _value!.load(ordering: .relaxed)
  }
}
```

To prevent such misuse, in the current implementation of this proposal, the code above generates a warning:

```text
warning: inout expression creates a temporary pointer, but argument 'at' should be 
a pointer that outlives the call to 'init(at:)'
    _value = UnsafeAtomic<Int>(at: &_storage)
                                   ^~~~~~~~~
```

This is implemented using a preexisting diagnostic based on a compiler heuristic. Ideally this warning would be promoted to a compile-time error.

> > **Note:** For an idea on how to add proper language support for taking the address of certain kinds of variables, see the discussion on the hypothetical `@raw` attribute in [Memory\-Safe Atomic Constructs](#memory-safe-atomic-constructs).

Reviewer: `@raw` ➡️ `@addressable`

```suggestion
> **Note:** For an idea on how to add proper language support for taking the address of certain kinds of variables, see the discussion on the hypothetical `@addressable` attribute in [Memory\-Safe Atomic Constructs](#memory-safe-atomic-constructs).
```

Without an efficient way to implement these two steps as a single atomic transaction, the implementation of `store` needs to delay releasing the overwritten value until it can guarantee that every outstanding load operation is completed. Exactly how to implement this is the problem of *memory reclamation* in concurrent data structures.

There are a variety of approaches to tackle this problem, some of which may be general enough to consider in future proposals. (One potential solution can be built on top of double-wide atomic operations, by offloading some of the reference counting operations into the second word of a double-wide atomic reference.)

(It'd be straightforward to use locks to build an atomic strong reference; while such a construct obviously wouldn't be lock-free, it is still a useful abstraction, so it may be a worthy addition to the Standard Library. However, locking constructs are outside the scope of this proposal.)

### Additional Low-Level Atomic Features

To enable usecases that require even more fine-grained control over atomic operations, it may be useful to introduce additional low-level atomics features:

> * support for additonal kinds of atomic values (such as double-wide atomics or floating-point atomics [[P0020]]),

Reviewer: `additonal` ➡️ `additional`

```suggestion
* support for additional kinds of atomic values (such as double-wide atomics or floating-point atomics [[P0020]]),
```
